### ISSUE TITLE
KMS: Key not granting permissions to admins anymore

### ISSUE THREAD

**Issue by icelava** [aws/aws-cdk#31528]
### Describe the bug

Previously with CDK 2.133, we were able to define additional IAM user ARNs for reference and adding them to a Key Admins array. They ended up in the key policy as expected.

```
var admins = this._platConfig.UploadsBucket.Admins.Select(u => User.FromUserArn(this, u, u));
var keyName = $"appuploads-key";
var uploadsKey = new Key(this, keyName, new KeyProps
	{
		Alias = keyName,
		Description = $"appuploads key.",
		Enabled = true,
		Admins = admins.ToArray()
	});
```

Now with 2.159.1 (build c66f4e3) it appears to ignore the array thus removing the previously defined users in the key policy. Even though the admins IEnumerable has IUser objects.

```
[~] AWS::KMS::Key appuploads-key appuploadskeyC9108645
 └─ [~] KeyPolicy
     └─ [~] .Statement:
         └─ @@ -6,28 +6,5 @@
            [ ]       "AWS": "arn:aws:iam::ACCOUNTID:root"
            [ ]     },
            [ ]     "Resource": "*"
            [-]   },
            [-]   {
            [-]     "Action": [
            [-]       "kms:CancelKeyDeletion",
            [-]       "kms:Create*",
            [-]       "kms:Delete*",
            [-]       "kms:Describe*",
            [-]       "kms:Disable*",
            [-]       "kms:Enable*",
            [-]       "kms:Get*",
            [-]       "kms:List*",
            [-]       "kms:Put*",
            [-]       "kms:Revoke*",
            [-]       "kms:ScheduleKeyDeletion",
            [-]       "kms:TagResource",
            [-]       "kms:UntagResource",
            [-]       "kms:Update*"
            [-]     ],
            [-]     "Effect": "Allow",
            [-]     "Principal": {
            [-]       "AWS": "arn:aws:iam::ACCOUNTID:user/old.user"
            [-]     },
            [-]     "Resource": "*"
            [ ]   }
            [ ] ]
```

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Last Known Working CDK Version

1.133

### Expected Behavior

The referenced IUsers added to Key.Admins array end up in key policy.

### Current Behavior

Referenced IUsers removed from key policy.

### Reproduction Steps

1. Update CDK CLI to 2.159.
2. cdk diff existing stack with KMS Key with additional IAM users as Admins.

### CDK CLI Version

2.159.1 (build c66f4e3)

### Node.js Version

v22.3.0

### OS

Windows 11 23H2

### Language

.NET

### Language Version

C# 10

**Comment by pahud** [comment]
investigating

**Comment by pahud** [comment]
Yes I noticed the changes between 2.141.0 and 2.142.0 and this could due to this fix https://github.com/aws/aws-cdk/pull/30023


Given the code below:

```ts
    const principal = iam.User.fromUserArn(this, 'User', `arn:aws:iam::${Stack.of(this).account}:user/foo`);
    new kms.Key(this, 'DummyKey', {
      admins: [principal]
    })
```

In 2.141.0, this would create both  IAM Policy for the User as well as the KMS key policy

Key Policy added:

```json
 {
            "Effect": "Allow",
            "Principal": {
                "AWS": "arn:aws:iam::ACCOUNT_ID:user/foo"
            },
            "Action": [
                "kms:CancelKeyDeletion",
                "kms:Create*",
                "kms:Delete*",
                "kms:Describe*",
                "kms:Disable*",
                "kms:Enable*",
                "kms:Get*",
                "kms:List*",
                "kms:Put*",
                "kms:Revoke*",
                "kms:ScheduleKeyDeletion",
                "kms:TagResource",
                "kms:UntagResource",
                "kms:Update*"
            ],
            "Resource": "*"
        }
```

In 2.142.0, this would ONLY add the IAM policy to the User. No additional key policy would be added.

Please note this is consist with the behavior of IAM Role with KMS Key. For example if I import an IAM role like this and deploy in 2.141.0

```ts
    const principal = iam.Role.fromRoleArn(this, 'Role', `arn:aws:iam::${Stack.of(this).account}:role/foo`);
    new kms.Key(this, 'DummyKey', {
      admins: [principal]
    })
```

You will notice no key policy would be added. Only IAM policy for the role would be added. If you upgrade from 2.141.0 to 2.142.0 or even the latest 2.159.1 and run `cdk diff`. There would be no differences.

So I am pretty sure this is a consistent behavior with what is happening for the kms key with an IAM role.

I will continue investigate and update here.

**Comment by pahud** [comment]
Hi @icelava 

```
var admins = this._platConfig.UploadsBucket.Admins.Select(u => User.FromUserArn(this, u, u));
```

Can you let me know, if the UserArn is from the same account of the deploying stack or different account?

**Comment by pahud** [comment]
OK I think I found the root cause.

Because of [this bug](https://github.com/aws/aws-cdk/issues/29999) in 2.141.0, which was fixed in https://github.com/aws/aws-cdk/pull/30023, when we import User from same account like

```ts
    const userPrincipal = iam.User.fromUserArn(this, 'User', `arn:aws:iam::123456789012:user/foo`);
```

CDK would always use `Aws.AccountId` to represents the grantee principal account and will always run this code logic

https://github.com/aws/aws-cdk/blob/3d1c06e5bf1fd128e81d6cf850130d4edc09ab02/packages/aws-cdk-lib/aws-kms/lib/key.ts#L184-L186

If we check [addToPrincipalOrResource()](https://github.com/aws/aws-cdk/blob/5e95ba2b7a24b2598cf00890e1e7d569914f27c9/packages/aws-cdk-lib/aws-iam/lib/grant.ts#L122), it first [addToPrincipal()](https://github.com/aws/aws-cdk/blob/5e95ba2b7a24b2598cf00890e1e7d569914f27c9/packages/aws-cdk-lib/aws-iam/lib/grant.ts#L166), which will succeed but `sameAccount` would be false here:

https://github.com/aws/aws-cdk/blob/5e95ba2b7a24b2598cf00890e1e7d569914f27c9/packages/aws-cdk-lib/aws-iam/lib/grant.ts#L141-L143

So it moves on to create resource policy here:

https://github.com/aws/aws-cdk/blob/5e95ba2b7a24b2598cf00890e1e7d569914f27c9/packages/aws-cdk-lib/aws-iam/lib/grant.ts#L145-L151

This explains why in 2.141.0 it would create both principal policy as well as resource policy.

In 2.142.0, because of the fix, `sameAccount` would be true so it won't move on to create the resource policy. This explain why it would be removed when upgrading from 2.141.0 to 2.142.0

**Comment by icelava** [comment]
> Can you let me know, if the UserArn is from the same account of the deploying stack or different account?

Yes the admin user is from the same AWS account.

**Comment by icelava** [comment]
So the key policy is redundant since the IAM user has its own granting policy?

**Comment by pahud** [comment]
When `admins` is defined, the Key construct essentially runs grantAdmin() on the principals which runs grant() on them with admin permissions. The whole implementation is [here](https://github.com/aws/aws-cdk/blob/218331bd1f87897bf6cbd42a2059d570b59bfb3f/packages/aws-cdk-lib/aws-kms/lib/key.ts#L164-L196).

https://github.com/aws/aws-cdk/blob/218331bd1f87897bf6cbd42a2059d570b59bfb3f/packages/aws-cdk-lib/aws-kms/lib/key.ts#L808

Basically, when CDK grants resource access permission to an IAM principal, if they both belong to the same account with the deploying stack, which I believe is your case, CDK would run addToPrincipalOrResource() which first add an IAM Policy for that identity principal and won't add resource-based policy if the previous add is successful. That explains why you'd see identity-based policy only but no resource-based policy.

And you are right, if you have multiple Key granting to the same principal, you would end up with multiple AWS::IAM::Policy resources being associated with a single IAM principal and that might not be something you want. In this case, I won't recommend using `admin` prop, instead, you probably need to create a single iam.Policy for that IAM principal and attach all necessary statements to that policy.

**Comment by icelava** [comment]
That will be difficult. Each application stack is not aware of each other and being self-contained assigns permissions to their respective administrators/developers. Anyway making the IUser construct ID unique across all stacks lets them maintain their own inine policies.

(Although the _2048 character limit_ on all inline policies is very restrictive.)
